src/Escher_Synth.ml

```diff
@@ -113,38 +114,54 @@ let solve_impl ?ast:(ast=false) task consts =
        List.iter ~f:(close_goal v) v_closes; true
   in
 
-  let int_components = List.filter ~f:(fun c -> c.codomain = TInt) components in
-  let bool_components = List.filter ~f:(fun c -> c.codomain = TBool) components in
-
   let apply_comp f types i =
     let rec apply_cells types acc locations = match types, locations with
-      | (typ::typs, i::locs) -> VSet.iter (fun x -> apply_cells typs (x::acc) locs) begin
-          match typ with
-            | TInt -> int_array.(i)
-            | TBool -> bool_array.(i)
-            | _ -> raise (Invalid_Type_Exn ("Escher does not handle type: "
+      | (typ::typs, i::locs) -> 
+      let apply_next = VSet.iter (fun x -> apply_cells typs (x::acc) locs) in begin
+        match typ with 
+          | TInt -> apply_next int_array.(i)
+          | TBool -> apply_next bool_array.(i)
+          | TList t -> apply_next list_array.(i)
+            (*
+            VSet.filter 
+            (fun (p, varray) -> (typeof varray.(0)) = (TList t)) list_array.(i) 
+            *)
+          | TVar s -> [int_array.(i); bool_array.(i); list_array.(i)] |> List.iter ~f:apply_next
+          | _ -> raise (Invalid_Type_Exn ("Escher does not handle type: "
                                            ^ (string_of_typ typ)))
-          end
+      end
       | ([], []) -> f (List.rev acc)
       | _ -> failwith "Impossible!"
     in divide (apply_cells types []) (List.length types) (i-1) []
   in
-  let expand_component c array i =
+  let expand_component c array_sel i =
     let f x =
+      (* We will modify c's codomain with the union result *)
+      try 
+      let typs = List.map ~f:(fun (_,x) -> typeof x.(0)) x in 
+      let codomain = 
+        let renamed_domain = rename (collect_varnames typs) c.domain in
+        let s = unify typs renamed_domain in apply s c.codomain in
+      let c = {c with codomain = codomain} in 
+      let array = array_sel c.codomain in 
       let vector = apply_component c x in
       let h_value = hvalue vector in
       let has_err = Array.fold ~f:(fun p x -> match x with VError -> true | _ -> p) ~init:false (snd vector) in
       if (h_value < !max_h && (not has_err))
       then ((if not (!noisy) then ()
             else print_endline (string_of_int h_value ^ ">>" ^ (Vector.string vector)));
             array.(h_value) <- VSet.add vector (array.(h_value)))
+      with _ -> ()
     in apply_comp f c.domain i
   in
-  let expand_type (mat, components) i =
-    List.iter ~f:(fun c -> expand_component c mat i) components;
-  in
   let expand i =
-    List.iter ~f:(fun x -> expand_type x i) [(int_array, int_components); (bool_array, bool_components)]
+    let array_sel t = match t with 
+    | TInt -> int_array 
+    | TBool -> bool_array
+    | TList _ -> list_array 
```

Comment:
Here TChar and TString is still pending. I'll add them later maybe, after all the logic is reviewed?